app/assets/javascripts/oxalis/model/sagas/isosurface_saga.js

```diff
@@ -0,0 +1,103 @@
+import { FlycamActions } from "oxalis/model/actions/flycam_actions";
+import { type Saga, _takeEvery, select, take } from "oxalis/model/sagas/effect-generators";
+import { V3 } from "libs/mjs";
+import type { Vector3 } from "oxalis/constants";
+import { computeIsosurface } from "admin/admin_rest_api";
+import { getFlooredPosition } from "oxalis/model/accessors/flycam_accessor";
+import { map3 } from "libs/utils";
+import DataLayer from "oxalis/model/data_layer";
+import Model from "oxalis/model";
+import getSceneController from "oxalis/controller/scene_controller_provider";
+
+class ThreeDMap<T> {
+  map: Map<number, ?Map<number, ?Map<number, T>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  get(vec: Vector3): ?T {
+    const [x, y, z] = vec;
+    if (this.map[x] == null) {
+      return null;
+    }
+    if (this.map[x][y] == null) {
+      return null;
+    }
+    if (this.map[x][y][z] == null) {
+      return null;
+    }
+    return this.map[x][y][z];
+  }
+
+  set(vec: Vector3, value: T): void {
+    const [x, y, z] = vec;
+    if (this.map[x] == null) {
+      this.map[x] = new Map();
+    }
+    if (this.map[x][y] == null) {
+      this.map[x][y] = new Map();
+    }
+    this.map[x][y][z] = value;
+  }
+}
+const isosurfacesMap: Map<number, ThreeDMap<boolean>> = new Map();
+const cubeSize = [256, 256, 256];
+
+function* ensureSuitableIsosurface(): Saga<void> {
+  const renderIsosurfaces = yield* select(state => state.datasetConfiguration.renderIsosurfaces);
+  if (!renderIsosurfaces) {
+    return;
+  }
+  const dataset = yield* select(state => state.dataset);
+  const layer = Model.getSegmentationLayer();
+  const position = yield* select(state => getFlooredPosition(state.flycam));
+  const segmentId = layer.cube.getDataValue(position, null, 1);
+
+  if (segmentId === 0 || segmentId == null) {
+    return;
+  }
+
+  if (isosurfacesMap.get(segmentId) == null) {
+    isosurfacesMap.set(segmentId, new ThreeDMap());
+  }
+  const threeDMap = isosurfacesMap.get(segmentId);
+
+  const zoomStep = 1;
+  const zoomedCubeSize = map3(el => el * 2 ** zoomStep, cubeSize);
+  const currentCube = map3((el, idx) => Math.floor(el / zoomedCubeSize[idx]), position);
+  const cubedPostion = map3((el, idx) => el * zoomedCubeSize[idx], currentCube);
+  if (threeDMap.get(currentCube)) {
+    return;
+  }
+
+  threeDMap.set(currentCube, true);
+  loadIsosurface(dataset, layer, segmentId, cubedPostion, zoomStep);
+}
+
+async function loadIsosurface(
+  dataset: APIDataset,
+  layer: DataLayer,
+  segmentId: number,
+  position: Vector3,
+  zoomStep: number,
+) {
+  const voxelDimensions = [2, 2, 2];
+
+  const responseBuffer = await computeIsosurface(
+    dataset,
+    layer,
+    V3.toArray(V3.sub(position, voxelDimensions)),
```

Comment:
I don't think I understand why we need to subtract the voxelDimensions here and also why we add it to the cubeSize. Could you briefly explain?

Reply:
The reason being, that we compute the mesh BETWEEN voxels and not AT voxels. So if the client requests 0-99, we will respond with all triangles between 0 and 1, 1 and 2, ... 98 and 99. If the follow-up request is 100-199, we are missing the triangles 99-100. The requests must thus be slightly overlapping